[Python] Add documentation of auto-instrumenting Python services using the operator (#2159)
open-telemetry/opentelemetry.io: content/en/docs/instrumentation/python/automatic/operator.md

---
title: Using the OpenTelemetry Operatory to Inject Auto-Instrumentation
linkTitle: Operator
weight: 45
spelling: cSpell:ignore distro mkdir uninstrumented virtualenv
---

If you run your Python service in Kubernetes, you can take advantage of the
[OpenTelemetry Operator](https://github.com/open-telemetry/opentelemetry-operator)
to inject auto-instrumentation without having to modify each of your services
directly.

> [See the OpenTelemetry Operator Auto-instrumentation docs for more details.](../../../../k8s-operator/automatic/)

Reviewer: ```suggestion
[See the OpenTelemetry Operator Auto-instrumentation docs for more details.](/docs/k8s-operator/automatic/)
```

---
title: Using the OpenTelemetry Operatory to Inject Auto-Instrumentation
linkTitle: Operator
weight: 45
spelling: cSpell:ignore distro mkdir uninstrumented virtualenv
---

> If running your python service in kubernetes you can take advantage of the [OpenTelemetry Operator](https://github.com/open-telemetry/opentelemetry-operator) to inject auto-instrumentation without having to make changes to your service's dependencies to run command.

Reviewer 2: ```suggestion
If you run your Python service in Kubernetes, you can take advantage of the [OpenTelemetry Operator](https://github.com/open-telemetry/opentelemetry-operator) to inject auto-instrumentation without having to modify each of your services directly.
```

## Setup

> First you must install the [OpenTelemetry Operator](https://github.com/open-telemetry/opentelemetry-operator) into your cluster.

Reviewer 2: ```suggestion
First, install the [OpenTelemetry Operator](https://github.com/open-telemetry/opentelemetry-operator) into your cluster.
```

> This can be done directly via the [Operator release manifest](https://github.com/open-telemetry/opentelemetry-operator#getting-started), via the [Operator helm chart](https://github.com/open-telemetry/opentelemetry-helm-charts/tree/main/charts/opentelemetry-operator#opentelemetry-operator-helm-chart), or via [Operator Hub](https://operatorhub.io/operator/opentelemetry-operator).  

Reviewer 2: ```suggestion
You can do this with the [Operator release manifest](https://github.com/open-telemetry/opentelemetry-operator#getting-started), the [Operator helm chart](https://github.com/open-telemetry/opentelemetry-helm-charts/tree/main/charts/opentelemetry-operator#opentelemetry-operator-helm-chart), or with [Operator Hub](https://operatorhub.io/operator/opentelemetry-operator).  
```

> In most cases you will need to the [cert-manager installed](https://cert-manager.io/docs/installation/), but if using the helm chart there is an option to generate a self-signed cert.

Reviewer 2: ```suggestion
In most cases, you will need to install [cert-manager](https://cert-manager.io/docs/installation/). If you use the helm chart, there is an option to generate a self-signed cert instead.
```

## Create an OpenTelemetry Collector (Optional)

> It is a best practice to send telemetry from containers to an [OpenTelemetry Collector](../../../../collector/) instead of directly to a backend.  The Collector helps simplify secret management, provides reliability and scalability, and can help improve data (k8sattributes processor).  If you chose not to use a Collector you can skip to the next section.

Reviewer 2: ```suggestion
It is a best practice to send telemetry from containers to an [OpenTelemetry Collector](../../../../collector/) instead of directly to a backend.  The Collector helps simplify secret management, decouples data export problems (such as a need to do retries) from your apps, and lets you add additional data to your telemetry, such as with the `k8sattributesprocessor` component. If you chose not to use a Collector, you can skip to the next section.
```

> The Operator provides a [Custom Resource Definition (CRD) for an OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#opentelemetrycollector) which is used to create an instance of the Collector that the Operator will manage. This example will deploy the Collector as a deployment, which is the default, but there are other [deployment modes](https://github.com/open-telemetry/opentelemetry-operator#deployment-modes) that can be used.

Reviewer 2: ```suggestion
The Operator provides a [Custom Resource Definition (CRD) for the OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#opentelemetrycollector) which is used to create an instance of the Collector that the Operator manages. The following example deploys the Collector as a deployment (the default), but there are other [deployment modes](https://github.com/open-telemetry/opentelemetry-operator#deployment-modes) that can be used.
```

```bash
kubectl apply -f - <<EOF
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: demo
spec:
  config: |
    receivers:
      otlp:
        protocols:
          grpc:
          http:
    processors:
      memory_limiter:
        check_interval: 1s
        limit_percentage: 75
        spike_limit_percentage: 15
      batch:
        send_batch_size: 10000
        timeout: 10s

    exporters:
      logging:

    service:
      pipelines:
        traces:
          receivers: [otlp]
          processors: [memory_limiter]
          exporters: [logging]
        metrics:
          receivers: [otlp]
          processors: [memory_limiter]
          exporters: [logging]
EOF
```

> This creates an deployment of the Collector that we can use as an endpoint for the auto-instrumentation.

Reviewer 2: ```suggestion
The above command results in a deployment of the Collector that you can use as an endpoint for auto-instrumentation in your pods.
```

> ## Configure the Instrumentation

Reviewer 2: ```suggestion
## Configure Autoinstrumentation
```

> In order to auto-instrument services the operator needs to be told how and what to instrument.  This is done via the [Instrumenation CRD](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#instrumentation).

Reviewer 2: ```suggestion
To be able to manage autoinstrumentation, the Operator needs to be configured to know what pods to instrument and which autoinstrumentation to use for those pods. This is done via the [Instrumentation CRD](https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#instrumentation).
```

> Creating the Instrumentation object correctly is paramount to getting auto-instrumentation working.  Via the instrumentation object you are  configuring the OpenTelemetry SDK for python. Making sure all endpoints and env vars are correct is required for auto-instrumentation to work properly.

Reviewer 2: I don't think I understand how this section relates to the code sample below. Where does it specify that we're using the Python SDK?

Author: This paragraph is supposed to be a warning that configuring the Instrumentation resource is both the trickiest but most important step of injecting auto-instrumentation.  I was trying to highlight that some of the items set in the Instrumentation resource map to items in the SDK that the operator configures and then inject.  

When I was learning this process I got held up for a second realizing that setting all the env vars in the Instrumentation object was the same as setting all the env var for an application that implemented the SDK itself.  In order to correctly configure the Instrumentation resource you have to understand the env vars that the SDK uses/depend on.

How does this rewording sound?

```suggestion
Creating the Instrumentation resource correctly is paramount to getting auto-instrumentation working.  Making sure all endpoints and env vars are correct is required for auto-instrumentation to work properly.
```

Reviewer 2: Yep, I like it.

```bash
kubectl apply -f - <<EOF
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
metadata:
  name: demo-instrumentation
spec:
  exporter:

>     endpoint: http://demo-collector:4318

Reviewer 2: Where did `demo-collector` come from? I see below that you wrote:

> which is the Collector's otlpreceiver's `http` port.

So this is the default name of the endpoint when you add the collector CRD?

Author: The OpenTelemetryCollector resourse automatically creates a service with the name of the OpenTelemetryCollector resource prepended to `-collector`.  Since our OpenTelemetryCollector resource was named `demo` we communicate with it as `demo-collector`.  I will call out that naming in the collector section.

Author: Updated

  propagators:
    - tracecontext
    - baggage
  sampler:
    type: parentbased_traceidratio
    argument: "1"
EOF
```

> By default the Instrumentation resource auto-instruments python services to export their telemetry using the `otlp exporter` with the `http/proto` protocol, which means the endpoint must be able to receive otlp `http/proto`.  Therefore, the example uses `http://demo-collector:4318`, which is the Collector's otlpreceiver's `http` port.

Reviewer 2: ```suggestion
By default, the Instrumentation resource that auto-instruments python services uses the `otlp` with the `http/proto` protocol. This means that the configured endpoint must be able to receive OTLP over `http/proto`.  Therefore, the example uses `http://demo-collector:4318`, which is the Collector's otlpreceiver's `http` port.
```

> As of operator v0.67.0 the `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL` and `OTEL_EXPORTER_OTLP_METRICS_PROTOCOL` env vars are automatically set to `http/proto`.  If you use an older version of the Operator you **MUST** set these env variables or python auto-instrumentation will not work.

Reviewer 2: ```suggestion
> As of operator v0.67.0, the `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL` and `OTEL_EXPORTER_OTLP_METRICS_PROTOCOL` env vars are automatically set to `http/proto`.  If you use an older version of the Operator you **MUST** set these env variables to `http/proto`, or python auto-instrumentation will not work.
```

> As of operator v0.67.0 the `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL` and `OTEL_EXPORTER_OTLP_METRICS_PROTOCOL` env vars are automatically set to `http/proto`.  If you use an old version of the Operator you **MUST** set these env variables or python auto-instrumentation will not work.

Reviewer 3: old --> older

> This document is working under the assumption that the only services opting into auto-instrumentation are python services.  If you have other languages to auto-instrument this instrumentation object will look different.  See the [Operators auto-instrumentation documentation](https://github.com/open-telemetry/opentelemetry-operator/blob/main/README.md#opentelemetry-auto-instrumentation-injection) for more details.

Reviewer 2: ```suggestion
```

Let's remove this for now -- we can add more stuff for other languages later (I'll create a tracking issue).

> At this point you have the ability to auto-instrument python services, but auto-instrumentation follows an opt-in model.  In order to become instrumented services must update their annotations.

Reviewer 2: ```suggestion
Now that your Instrumentation object is created, your cluster has the ability to auto-instrument python services and send data to an endpoint. However, auto-instrumentation with the OpenTelemetry Operator follows an opt-in model. In order to activate autoinstrumentation, you'll need to add an annotation to your deployment.
```

## Add annotations to existing deployments

> The final step is to opt in your python services to auto-instrumentation.  This is done by updating the service’s `spec.template.metadata.annotations` to include `instrumentation.opentelemetry.io/inject-python: "true"`.  Alternatively, the annotation can be added to the namespace and all services in the namespace will opt-in.   See the [Operators auto-instrumentation documentation](https://github.com/open-telemetry/opentelemetry-operator/blob/main/README.md#opentelemetry-auto-instrumentation-injection) for more details.

Reviewer 2: ```suggestion
The final step is to opt in your python services to autoinstrumentation.  This is done by updating your service’s `spec.template.metadata.annotations` to include `instrumentation.opentelemetry.io/inject-python: "true"`.

Alternatively, the annotation can be added to a namespace, which will result in all services in that namespace to opt-in to autoinstrumentation. See the [Operators auto-instrumentation documentation](https://github.com/open-telemetry/opentelemetry-operator/blob/main/README.md#opentelemetry-auto-instrumentation-injection) for more details.
```

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-flask-example
  labels:
    app.kubernetes.io/name: flask-example
spec:
  ...
  template:
    metadata:
      labels:
        app.kubernetes.io/name: flask-example
      annotations:
        instrumentation.opentelemetry.io/inject-python: "true"
    ...
```

> 
> 
> 
> 
> 

Reviewer 2: ```suggestion

```
